carp.common/src/commonMain/kotlin/dk/cachet/carp/common/application/URL.kt

```diff
@@ -0,0 +1,101 @@
+package dk.cachet.carp.common.application
+
+import dk.cachet.carp.common.infrastructure.serialization.createCarpStringPrimitiveSerializer
+import kotlinx.serialization.KSerializer
+import kotlinx.serialization.Serializable
+import kotlin.js.JsExport
+
+/**
+ * Represents a Uniform Resource Locator (URL) according to RFC 3986. We use [URIRegex] to match
+ * [stringRepresentation], but it cannot do actual validation. Therefore, [DefaultURLValidator] is used to validate
+ * the URL.
+ *
+ * The URL of the web page which contains the task to be performed.
```

Comment:
But, the time when the variable is represented and when it is translated to an actual URL, is different. For a `WebTask`, the variable placeholder is represented as such in the study protocol, and replaced by the client runtime when the task is triggered.

For device registrations, you may want to store a URL on the studies subsystem front-end somewhere, and after each invitation construct a URL to be used in the `WebBrowserDeviceRegistration` when pre-registering the device. But the variable placeholder is never stored in any of the currently designed CARP objects.

So, if you want to extract anything, it would probably be some type of `ParameterizedWebUrl`, which could encapsulate the `markup` functionality, and associated URL + variable validation, based on the "expected" variables.

Reply:
I think I want to refrain from doing something like a `ParameterizedWebUrl`. I guess I will just use `URL` to have a validated url and nothing else. And that will also avoid issues with things like the following.

> If it would actually be a regex to validate URLs, it would fail for expected input to this class, namely the URL variables. They are formatted as {{VAR}}, and { and } are not valid characters, which is exactly why they are a good pick for injecting variables. :)

Reply:
If you want to validate, you'd probably still want to restrict it to a WebUrl (http/https) though. I have a hard time imagining what a client running on ftp would look like. :) And, since you link it to web browser requirements.

Reply:
I think the requirements are not clear for now, so I dissected this pr into two, and made a draft ( #466 ) of the URL class. Maybe we could turn that into an issue. I make the registration take a string for now.